src/keywrap/pkcs11/mod.rs

```diff
@@ -0,0 +1,333 @@
+                }
+            },
+        }
+        pkcs11_keys.push(k);
+    }
+
+    Ok(pkcs11_keys)
+}
+
+
+
+
+#[cfg(test)]
+mod kw_tests {
+    use super::*;
+
+    //const SOFTHSM_SETUP: &str = "../../scripts/softhsm_setup";
+
+    #[test]
+    fn test_keywrap_pkcs11_success() {
+        let path_to_script = env!("CARGO_MANIFEST_DIR").to_string() + "/scripts/softhsm_setup";
+        let vs = create_valid_pkcs11_ccs().unwrap();
+        let valid_pkcs11_ccs = vs.0;
+        let shsm = vs.1;
+
+        // FIXME
+        //std::env::set_var("OCICRYPT_OAEP_HASHALG", "sha1");
+        std::env::set_var("OCICRYPT_OAEP_HASHALG", "sha256");
+
+        for cc in valid_pkcs11_ccs {
+            let kw = Pkcs11KeyWrapper{};
+
+            let data = "This is some secret text".as_bytes();
+
+            if let Some(ec) = cc.encrypt_config {
+                let wk = kw.wrap_keys(&ec, data).unwrap();
+                if let Some(dc) = cc.decrypt_config {
+                    let ud = kw.unwrap_keys(&dc, &wk).unwrap();
+                    assert_eq!(data, ud);
+                } else {
+                    assert!(false);
+                }
+            } else {
+                assert!(false);
+            }
+
+
+        }
+
+        shsm.run_softhsm_teardown(&path_to_script);
+    }
+
+    #[test]
+    fn test_annotation_id() {
+        let pkcs11_key_wrapper = Pkcs11KeyWrapper{};
+        assert_eq!(pkcs11_key_wrapper.annotation_id(),
+                   "org.opencontainers.image.enc.keys.pkcs11");
+    }
+
+    #[test]
+    fn test_no_possible_keys() {
+        let pkcs11_key_wrapper = Pkcs11KeyWrapper{};
+        let dc = DecryptConfig::default();
+        assert!(pkcs11_key_wrapper.no_possible_keys(&dc.param));
+    }
+
+    #[test]
+    fn test_private_keys() {
+        let pkcs11_key_wrapper = Pkcs11KeyWrapper{};
+        let dc = DecryptConfig::default();
+        assert!(pkcs11_key_wrapper.private_keys(&dc.param).is_none());
+        // TODO: test positive case (is_some)
+    }
+
+    #[test]
+    fn test_key_ids_from_packet() {
+        let pkcs11_key_wrapper = Pkcs11KeyWrapper{};
+        assert!(pkcs11_key_wrapper.keyids_from_packet("".to_string()) == None);
+    }
+
+    #[test]
+    fn test_recipients() {
+        let pkcs11_key_wrapper = Pkcs11KeyWrapper{};
+        let recipients = pkcs11_key_wrapper.recipients("".to_string()).unwrap();
+        assert!(recipients.len() == 1);
+        assert!(recipients[0] == "[pkcs11]");
+    }
+
+    fn load_data_path() -> String {
+        let mut path = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        path.push("data");
+        path.to_str().unwrap().to_string()
+    }
+
+
+    fn get_pkcs11_config_yaml() -> Result<Vec<u8>> {
+        // we need to provide a configuration file so that on the various
+        // distros the libsofthsm2.so will be found by searching directories
+        let mdyaml = get_default_module_directories_yaml("".to_string())?;
```

Comment:
similar to previous comments: create the config object, and then use serde to serialize as needed